rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,170 @@
+# RFC 8288 - 2021-07-20 - CSV Enrichment
+
+Vector needs to allow users to enrich the events flowing through the topology
+using through the topology using a CSV file. This RFC proposes adding a Vrl
+function that can perform this lookup.
+
+## Scope
+
+The RFC will cover a simple VRL function that will lookup a single row from a
+CSV file. 
+
+The lookup will be a simple sequential scan. The first version of this
+function will not handle any indexing of this data. Above a certain size
+indexing will provide substantial performance benefits, so it is something that
+we should definitely plan for implementing at a future stage. The RFC will 
+briefly discuss this, but will not go into any specific indexing strategies.
+
+## Motivation
+
+Users would like to enrich events flowing through Vector with extra data 
+provided by a CSV file. 
+
+## Doc-level Proposal
+
+Two remap functions:
+
+### `lookup_csv_row`
+
+This function will look up a single row within the dataset. If a single row is 
+found that data is returned as an object, otherwise this function will error. 
+
+### `lookup_csv_rows` 
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the 
+array.
+
+### Parameters 
+
+*filename*
+
+The filename is the path to the filename of the csv file to lookup. This
+filename has to be a static string (cannot be changed at runtime) so that 
+Vector can:
+
+1. Validate that the file exists.
+2. Load the file into memory.
+
+The filename will be relative to the current working directory that Vector is
+running in.
+
+The CSV file must start with a title row that gives a name to each of the 
+columns provided in the file.
+
+*key*
+
+The CSV file may be encrypted. This parameter specifies a key to use to decrypt
+the file.
+
+Is it safe to assume/require the file to be encrypted using AES, or should we
+support other algorithms as well?
```

Comment:
This is a tricky one.

I think it's fine to assume this for a v1, but this doesn't scale well, especially given CSV is but one way to store data, and we might want to introduce other functions in the future that are basically the same, but read a JSON file, or any other data file.

This _might_ even be an even stronger reason to move file access outside VRL.

For example, you could do something like:

```toml
[transforms.remap]
type = "remap"
whitelist_paths = [
	{ name = "mycsv", decrypt = { ... }, path = "/opt/mycsvs/my.csv" }
]
source = """
	lookup_csv_row(.data, p',mycsv')
"""
```

Basically, this would be an extension to what I mentioned previously:

- You _have_ to specify individual files you want to access in Vector's config, it reads the files at boot time, and provides a reference to the content to VRL's compiler when it is initialized.
- The name you give it has to be unique, and is used to reference the file inside VRL
- VRL has a `p'<unique name>'` literal that maps the name to the data belonging to that file
- Because VRL receives a reference to the decrypted file contents, we don't have to worry about any function arguments that have to do with decryption.

Reply:
or even a global setting like:

```
cache.csv.custom_data = "/data.csv"
```

Reply:
Agreed, the decryption bit especially seems like it would motivate a separate "load" step.

If we go that route, we could use `f'mycsv'` to indicate it is a file (rather than `p` for "path").

Another way this could be modeled is like:

```
csv = open_csv("foo.csv", ...)
lookup_csv_row(csv, ...)
```

But this would be a bit of a first in VRL to create objects that are used later so I think I prefer the separate configuration in the `remap` transform. The alternative also gives the impression that the file is loaded every time.

However, modeling it as `remap` configuration also means that this VRL feature would be `remap` specific and not usable in other places VRL is used (like `filter`, `route` and unit tests) unless we add this configuration there too.

Reply:
> However, modeling it as `remap` configuration also means that this VRL feature would be `remap` specific and not usable in other places VRL is used (like `filter`, `route` and unit tests) unless we add this configuration there too.

Correct, which is why I think we want to lift this into a top-level domain at some point. For now, we could add a component option that you can easily copy/paste to other components, similar to our existing `TlsConfig`, `BatchConfig` and the like.

We would also update the REPL to allow repeated use of arguments such as `--file mycsv=/foo/my.csv` to act similar to Vector's config field. For now, I'm assuming we can skip decrypting files in the REPL, but I imagine we can add `--file-decrypt-aes mycsv=$KEY` to specifically force decrypting provided files.

> If we go that route, we could use `f'mycsv'` to indicate it is a file (rather than `p` for "path").

Agreed, that does seem better if we go with the "reference the exact file" route.

Reply:
I'm really liking the top level configuration option.

Another consideration is that we need to allow for refreshing the file contents. It would not be a good solution to force people to have to restart Vector if they want to update the file. If we make it a top level option it would fit better with either allowing it to refresh the file on a SIGHUP, or perhaps watching the file for modification time changes, or even inotify.

When we add additional data sources, some of those may need to fit into the async ecosystem. Keeping the data source outside or VRL alleviates VRL from having to fit into that, keeping the core functionality simpler.

Downside, it does make the upfront development of this feature more complicated.

Reply:
I'd note that @blt was recently talking about the removal of config watching, I believe in the pipelines RFC

Reply:
Interesting, yeah, I'm down to entertain that. For now, we should probably aim for this to integrate with the existing `SIGHUP` reloading.

Reply:
If I may jump in here 😄 . I do not think we should concern ourselves with encryption or reloading at this stage. Let's just get a simple static CSV file working, we can follow up with those enhancements.

Reply:
> I do not think we should concern ourselves with encryption

I was somewhat believing that encryption was a requirement put forward by the people who requested this function. If not, then I fully agree that we should move encryption to a follow-up RFC whenever that becomes relevant.

Reply:
Yes, that's my bad. I wasn't aware this was a requirement in the brief. Even so, I think we'd benefit from deferring it on this first version and focusing on the core functionality. I agree that we should consider it for our design, ensuring that we can accomodate it easily, but not solve it right now.

Reply:
The proposal I made w.r.t. moving file reading outside VRL is almost entirely driven by the brief's requirement of decryption support.

The problem is that this is going to affect every file-based function we're adding, and I fully expect there to be more, other than the CSV one this RFC proposes. This is in addition to the complication of having to special-case path strings in the proposal you made w.r.t. having fully qualified paths be passed into the function as an argument.

Whatever way you slice or dice it, requiring local disk access _and_ having to support file decryption is going to add complications, either inside VRL, or out. My reasoning for moving it out of VRL is because it allows us more flexibility on how we implement it, and reduces the complicated matrix of parameters we're going to have to accept going forward (think: providing the file path itself, in addition to supporting a myriad of encryption schemes and their different options (which was well explained by @bruceg  in https://github.com/timberio/vector/issues/4870#issuecomment-721997201).

I think it's worth fleshing out both ways of solving this (entirely inside VRL, or through Vector's config) and list the pro's and con's, before we make a final decision on this.